### PR TITLE
Fix hdf.DataSource and improve grid handling

### LINKED ISSUE #83: 🐛[BUG]: era5.HDF5Source broken
### Version

main

### On which installation method(s) does this occur?

_No response_

### Describe the issue

- AttributeError: 'HDF5DataSource' object has no attribute 'grid'
- shape of output is (h, c, y, x) but should be (c, y, x)

we should add coverage for these use case. mypy should be able to catch the first (#52). the second requires a test to assert the shape is as expected at run time.

### Environment details

_No response_

### PR DESCRIPTION
**Pull request by nbren12**
# Earth-2 MIP Pull Request

## Description

Closes #83. 
- adds tests of hdf5 DataSource
- renames to hdf5.DataSource
- Makes a separate earth2mip.grid.LatLonGrid object for python API use and decouple the implementations from the configuration object `schema.Grid`.

## Checklist

- [x] I am familiar with the Contributing Guidelines.
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The CHANGELOG.md is up to date with these changes.
- [x] An [issue](https://github.com/NVIDIA/earth2mip/issues) is linked to this pull request.

### CODE REVIEW

File: tests/test_initial_conditions.py [hunk 1]
```diff
@@ -1,57 +0,0 @@
-# SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
```
Comment (nbren12) [inline tests/test_initial_conditions.py:1]: combined with tests/initial_conditions/test_initial_conditions and deleted redundant tests.

File: earth2mip/regrid.py [hunk 2]
```diff
@@ -59,53 +57,29 @@ def forward(self, x):
...
     if src == dest:
         return Identity()
-    elif (src, dest) == (Grid.grid_721x1440, Grid.grid_720x1440):
-        return DropSouthPole()
-    else:
-        return _get_tempest_regridder(src, dest)
-    raise NotImplementedError()
-
-
-def xarray_regrid(src: xarray.Dataset, dest: Grid):
```
Comment (nbren12) [inline earth2mip/regrid.py:88]: not used anywhere.

**Comment by nbren12** [comment]
I tested some examples manually:

- inference_ensemble with this config:
```
{
    "ensemble_members": 4,
    "noise_amplitude": 0.05,
    "simulation_length": 10,
    "weather_event": {
        "properties": {
            "name": "Globe",
            "start_time": "2018-01-01 00:00:00",
            "initial_condition_source": "cds"
        },
        "domains": [
            {
                "name": "global",
                "type": "Window",
                "diagnostics": [
                    {
                        "type": "raw",
                        "channels": [
                            "t2m"
			]
                    }
                ]
            }
        ]
    },
    "output_path": "outputs/graphcast",
    "output_frequency": 1,
    "weather_model": "e2mip://dlwp",
    "seed": 12345,
    "use_cuda_graphs": false,
    "ensemble_batch_size": 1,
    "autocast_fp16": false,
    "perturbation_strategy": "correlated",
    "noise_reddening": 2.0
}
```
- python3 examples/workflows/dlwp.py

File: earth2mip/grid.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+from typing import List
+import numpy as np
+import dataclasses
+from earth2mip import schema
+
+
+@dataclasses.dataclass(frozen=True)
+class LatLonGrid:
+
+    lat: List[float]
```
Comment (NickGeneva) [inline earth2mip/grid.py:25]: I see the merit of using python generics here, I feel most of the time this is used for initializing a xarray dataset where a numpy array is needed. So having these be numpy arrays feels like its more practical?

File: earth2mip/grid.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import List
+import numpy as np
+import dataclasses
+from earth2mip import schema
```
Comment (NickGeneva) [inline earth2mip/grid.py:19]: Could move the grid schema over here so all grid stuff is in one spot. The schema is essentially just known grids.

File: earth2mip/grid.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+    lat: List[float]
+    lon: List[float]
+
+    @property
+    def shape(self):
+        return (len(self.lat), len(self.lon))
+
+
+def regular_lat_lon_grid(
```
Comment (NickGeneva) [inline earth2mip/grid.py:33]: Nit: if this is regular lat lon grid, can we just call it `lat_lon_grid` and anything different is then `something_lat_lon_grid`

Seems a little too verbose

File: earth2mip/initial_conditions/hdf5.py [hunk 6]
```diff
@@ -26,39 +26,49 @@
...
-        if self.n_history != 0:
-            warnings.warn(
-                DeprecationWarning(
-                    "n_history is deprecated. Use earth2mip.initial_conditions.get_initial_condition_for_model instead."  # noqa
-                )
-            )
 
     @classmethod
-    def from_path(cls, root: str, **kwargs: Any) -> "HDF5DataSource":
+    def from_path(cls, root: str, **kwargs: Any) -> "DataSource":
```
Comment (NickGeneva) [inline earth2mip/initial_conditions/hdf5.py:59]: Why the quotes?

File: earth2mip/networks/fcn.py [hunk 7]
```diff
@@ -63,9 +63,8 @@ def load(package, *, pretrained=True, device="cuda"):
 
     core_model = modulus.Module.from_checkpoint(package.get("fcn.mdlus"))
 
-    grid = schema.Grid.grid_720x1440
     dt = datetime.timedelta(hours=6)
-
+    grid = earth2mip.grid.regular_lat_lon_grid(721, 1440)
```
Comment (NickGeneva) [inline earth2mip/networks/fcn.py:67]: 720 endpoint false

File: earth2mip/grid.py (hunk 5 above)
Comment (nbren12) [reply earth2mip/grid.py:33]: rename to equiangular_lat_lon?

**Review by NickGeneva — APPROVED** [review]

File: earth2mip/grid.py (hunk 5 above)
Comment (nbren12) [reply earth2mip/grid.py:33]: done.

File: earth2mip/networks/fcn.py (hunk 7 above)
Comment (nbren12) [reply earth2mip/networks/fcn.py]: done

**Pull request merged**